src/fx/TimeStretchPitchScaleReader.cpp

```diff
@@ -0,0 +1,285 @@
+/*******************************************************************************
+ * Copyright 2009-2025 Jörg Müller
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+#include "fx/TimeStretchPitchScaleReader.h"
+
+#include <iostream>
+
+#include "IReader.h"
+
+#include "util/Buffer.h"
+
+using namespace RubberBand;
+
+AUD_NAMESPACE_BEGIN
+
+TimeStretchPitchScaleReader::TimeStretchPitchScaleReader(std::shared_ptr<IReader> reader, double timeRatio, double pitchScale, StretcherQualityOptions quality) :
+    EffectReader(reader),
+    m_timeRatio(timeRatio),
+    m_pitchScale(pitchScale),
+    m_position(0),
+    m_length(0),
+    m_quality(-1),
+    m_finishedReader(false),
+    m_input(reader->getSpecs().channels),
+    m_processData(reader->getSpecs().channels),
+    m_output(reader->getSpecs().channels),
+    m_retrieveData(reader->getSpecs().channels)
+{
+	configure(quality);
+}
+
+void TimeStretchPitchScaleReader::read(int& length, bool& eos, sample_t* buffer)
+{
+	if(length == 0)
+		return;
+
+	if(m_pitchScale == 1.0 && m_timeRatio == 1.0)
+	{
+		m_reader->read(length, eos, buffer);
+		return;
+	}
+
+	int samplesize = AUD_SAMPLE_SIZE(m_reader->getSpecs());
+
+	int channels = m_reader->getSpecs().channels;
+	int len;
+	sample_t* buf;
+
+	int available = m_stretcher->available();
+	while(available < length + m_dropAmount && !m_finishedReader)
+	{
+		// size_t need = m_stretcher->getSamplesRequired();
+		// if (need == 0) break;
+
+		// Note for the V3 engine, the needed samples can be 0 sometimes and it breaks out of the loop too early before getting the necessary length for the buffer
+		// For now, choose the block size process size to be 1024.
+		// It's also actually faster too than to always use getSamplesRequired();
+
+		len = 1024;
```

Comment:
I think the number of samples you use here should depend on the length you need to output and the time stretch parameter. I.e., if you speed up time 2x, it should need 2x more input samples than you want output samples. Of course that's not a precise number which is why this needs to be a loop and you have to add more samples as necessary.